### PR TITLE
45 redesign sql statement factory

### PR DESCRIPTION
**Pull request by DrDiHai**

### CODE REVIEW

File: backend/src/db/sqlite.py [hunk 1]
```diff
@@ -17,13 +18,37 @@
...
+        if type in self.type_map:
+            self.data_type = self.type_map[type]
+            self.name = name
+        else:
+            raise ValueError(f"Unsupported data type for an SQLite Column definition: {type}")
+
+class SQLiteSQLFactory(SQLFactory):
+
+    def getClass(self, sql_cls: type):
+        if sql_cls.__name__ == "SQL_column_definition":
```
Comment (hk-ret) [inline backend/src/db/sqlite.py:40]: Namen Attribut der Basisklasse verwenden: `sql_cls.__name__ in [b.__name__ for b in SQLiteColumnDefinition.__bases__ ]`

File: backend/src/db/db_base.py [hunk 2]
```diff
@@ -14,6 +15,8 @@ def __init__(self, **cfg) -> None:
         self._cfg = cfg
         self._connections = set()
 
+    sqlFactory = SQLFactory
```
Comment (hk-ret) [inline backend/src/db/db_base.py:18]: it's probably clearer to implement `sqlFactory` as property to overwritten by the derived class.

File: backend/src/db/sql_statement.py [hunk 3]
```diff
@@ -0,0 +1,427 @@
...
+            ],
+        ).execute()"""
+
+    def sql(self)->str:
+        if self.sql_statement is None:
+            raise InvalidSQLStatementException("No SQL statement to execute.")
+        return self.sql_statement.sql()
+
+    rslt = None
+    sql_statement = None
```
Comment (hk-ret) [inline backend/src/db/sql_statement.py:60]: these are in fact instance attributes and should be initialized in `__init__()` and not declared as class attributes:

File: backend/src/persistance/business_object_base.py [hunk 4]
```diff
@@ -97,54 +107,37 @@ async def store(self):
...
-                    commit=True,
-                )
-            finally:
-                await self.fetch()
+            self._update_self()
+
+    async def _insert_self(self):
+        assert self.id is None, "id must be None for insert operation"
+        sql = SQL(App.db)
+        insert: Insert = sql.insert(self.table)
```
Comment (hk-ret) [inline backend/src/persistance/business_object_base.py:118]: why is there an `sql` and an `insert` object?

File: backend/testing/db/test_sql_statement.py
Comment (hk-ret) [inline backend/testing/db/test_sql_statement.py:1]: not reviewed, will do that another time

**Comment by hk-ret** [comment]
reviewed and found OK.

**Comment by DrDiHai** [comment]
Danke, werde den main in meinem Branch merheb.

hk-ret ***@***.***> schrieb am Mo., 10. Juni 2024, 17:09:

**Pull request merged**